solutions/cses-1163.mdx

```diff
@@ -61,47 +62,151 @@ int main(){
+output array. Then, we'll remove traffic lights in the reverse order to how they
+were added, and find the gap each removal creates.
+
+This gap is just the distance between the two street coordinates (either a
+traffic light or the beginning or ending of the street) next to a given traffic
+light stored in our set, so we can use our set to find these values and subtract
+them to get our gap.
+
+But this gap may not be the maximum gap. We'll compare this to the gap we found
+once all the traffic lights are added, and set the max gap to the greater value.
+We'll then set the next lowest element in the output array to this value, which
+will represent the greatest gap prior to adding the traffic light we just
+removed.
+	
+## Implementation
+**Time Complexity:** $\mathcal{O}(n)$
+	
+<LanguageSection>
+<JavaSection>
+
+<Warning>
+Due to the small CSES time limit, it's possible that standard input/output will occasionally TLE.
+</Warning>
+```java
+import java.io.*;
+import java.util.*;
+
+public class TrafficLights {
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio(); 
+	
+		int streetLength = io.nextInt();
+		int lightNum = io.nextInt();
+
+		// Using an array to read values since we can't get values from sets in Java
+		int[] opArray = new int[lightNum];
+		NavigableSet<Integer> streetPositions = new TreeSet<>();
+		// Initialize the set with beginning and ending values
+		streetPositions.add(0);
+		streetPositions.add(streetLength);
+
+		for (int i = 0; i < lightNum; i++) {
+			int nextTrafficLight = io.nextInt();
+			opArray[i] = nextTrafficLight;
+			streetPositions.add(nextTrafficLight);
+		}
+		int[] gapsArray = new int[lightNum];
+		int prev = 0;
+		int maxGap = 0;
+		// Find the longest passage once all the streetlights are added
+		for (int i : streetPositions) {
+			maxGap = Math.max(i - prev, maxGap);
+			prev = i;
+		}
+		gapsArray[lightNum - 1] = maxGap;
+		/*
+		 * Remove the streetlights in reverse order to how they were added, then find
+		 * the gap created by removing each. Find the biggest current gap, and 
+		 * add it to the next lowest index in answer.
+		 */
+		for (int i = lightNum - 1; i > 0; i--) {
+			streetPositions.remove(opArray[i]);
+			int low = streetPositions.lower(opArray[i]);
+			int high = streetPositions.higher(opArray[i]);
+			maxGap = Math.max(maxGap, high - low);
+			gapsArray[i - 1] = maxGap;
+		}
+		//  Use StringBuilder to print out the array quicker
+		StringBuilder sb = new StringBuilder();
+		for (int i : gapsArray) {
+			sb.append(i).append(" ");
+		}
+		io.println(sb);
+		io.close();
+	}
+	CodeSnip{Kattio}
+}
+	
+```
+	
+</JavaSection>
+<CPPSection>
+	
 ```cpp
-#include <bits/stdc++.h>
-
-using namespace std;
-
-//constant initialization
-const int maxn=2e5+10;
-
-//variables used for the current problem
-int x,n,p[maxn],mx; set<int> st; vector<int> ans;
+#include <iostream>
+#include <vector>
+#include <set>
+#include <algorithm>
```

Comment:
```suggestion
#include <bits/stdc++.h>
```

Reply:
hey that's my code!
uh best practice, mods said they didn't really care